### PR TITLE
grass: pin to python 3.11

### PR DESCRIPTION
**Pull request by fpletz**
## Description of changes

Pin to Python 3.11 since wxpython is not yet compatible to Python 3.12.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Review by mjoerg — COMMENTED** [review]
Result of `nixpkgs-review pr 325731` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package failed to build:</summary>
  <ul>
    <li>qgis-ltr</li>
  </ul>
</details>
<details>
  <summary>2 packages built:</summary>
  <ul>
    <li>grass</li>
    <li>qgis</li>
  </ul>
</details>

**Comment by fpletz** [comment]
Thanks, fixed qgis-ltr.

**Comment by fpletz** [comment]
qgis-ltr still fails. But this is not due to the update of the python version but the update of gqgis to 3.34.8: https://hydra.nixos.org/job/nixpkgs/trunk/qgis-ltr.x86_64-linux. Thus unrelated to this PR.

**Comment by imincik** [comment]
I [QGIS is building successfully](https://hydra.nixos.org/job/nixpkgs/trunk/qgis.x86_64-linux/all) even without this PR on Hydra.

**Comment by imincik** [comment]
I am not sure why QGIS-LTR fails. It was building successfully at the [moment of 3.34.8](https://github.com/NixOS/nixpkgs/pull/321562/checks?check_run_id=26592720914) update.

File: pkgs/applications/gis/grass/default.nix [hunk 1]
```diff
@@ -23,7 +23,7 @@
 , pkg-config
 , postgresql
 , proj
-, python3Packages
```
Comment (imincik) [inline pkgs/applications/gis/grass/default.nix:26]: May I suggest to set python packages in let binding so we don't repeat python version over and over.

```diff
diff --git a/pkgs/applications/gis/grass/default.nix b/pkgs/applications/gis/grass/default.nix
index 4eb10a34d2e2..e6037c4674e6 100644
--- a/pkgs/applications/gis/grass/default.nix
+++ b/pkgs/applications/gis/grass/default.nix
@@ -23,7 +23,7 @@
 , pkg-config
 , postgresql
 , proj
-, python3Packages
+, python311Packages
 , readline
 , sqlite
 , wxGTK32
@@ -31,6 +31,10 @@
 , zstd
 }:

+let
+  pyPackages = python311Packages;
+
+in
 stdenv.mkDerivation (finalAttrs: {
   pname = "grass";
   version = "8.3.2";
@@ -53,7 +57,7 @@ stdenv.mkDerivation (finalAttrs: {
     libmysqlclient # for `mysql_config`
     netcdf # for `nc-config`
     pkg-config
-  ] ++ (with python3Packages; [ python-dateutil numpy wxpython ]);
+  ] ++ (with pyPackages; [ python-dateutil numpy wxpython ]);

   buildInputs = [
     blas
@@ -86,7 +90,7 @@ stdenv.mkDerivation (finalAttrs: {

   # Correct mysql_config query
   postPatch = ''
-      substituteInPlace configure --replace "--libmysqld-libs" "--libs"
+    substituteInPlace configure --replace "--libmysqld-libs" "--libs"
   '';

   configureFlags = [
@@ -127,7 +131,7 @@ stdenv.mkDerivation (finalAttrs: {
   postInstall = ''
     wrapProgram $out/bin/grass \
     --set PYTHONPATH $PYTHONPATH \
-    --set GRASS_PYTHON ${python3Packages.python.interpreter} \
+    --set GRASS_PYTHON ${pyPackages.python.interpreter} \
     --suffix LD_LIBRARY_PATH ':' '${gdal}/lib'
     ln -s $out/grass*/lib $out/lib
     ln -s $out/grass*/include $out/include
```

Comment (imincik) [inline pkgs/applications/gis/grass/default.nix:26]: May I suggest to set python packages in let binding so we are not repeating python version multiple times ?

```diff
diff --git a/pkgs/applications/gis/grass/default.nix b/pkgs/applications/gis/grass/default.nix
index 4eb10a34d2e2..e6037c4674e6 100644
--- a/pkgs/applications/gis/grass/default.nix
+++ b/pkgs/applications/gis/grass/default.nix
@@ -23,7 +23,7 @@
 , pkg-config
 , postgresql
 , proj
-, python3Packages
+, python311Packages
 , readline
 , sqlite
 , wxGTK32
@@ -31,6 +31,10 @@
 , zstd
 }:

+let
+  pyPackages = python311Packages;
+
+in
 stdenv.mkDerivation (finalAttrs: {
   pname = "grass";
   version = "8.3.2";
@@ -53,7 +57,7 @@ stdenv.mkDerivation (finalAttrs: {
     libmysqlclient # for `mysql_config`
     netcdf # for `nc-config`
     pkg-config
-  ] ++ (with python3Packages; [ python-dateutil numpy wxpython ]);
+  ] ++ (with pyPackages; [ python-dateutil numpy wxpython ]);

   buildInputs = [
     blas
@@ -86,7 +90,7 @@ stdenv.mkDerivation (finalAttrs: {

   # Correct mysql_config query
   postPatch = ''
-      substituteInPlace configure --replace "--libmysqld-libs" "--libs"
+    substituteInPlace configure --replace "--libmysqld-libs" "--libs"
   '';

   configureFlags = [
@@ -127,7 +131,7 @@ stdenv.mkDerivation (finalAttrs: {
   postInstall = ''
     wrapProgram $out/bin/grass \
     --set PYTHONPATH $PYTHONPATH \
-    --set GRASS_PYTHON ${python3Packages.python.interpreter} \
+    --set GRASS_PYTHON ${pyPackages.python.interpreter} \
     --suffix LD_LIBRARY_PATH ':' '${gdal}/lib'
     ln -s $out/grass*/lib $out/lib
     ln -s $out/grass*/include $out/include
```

File: pkgs/applications/gis/qgis/unwrapped-ltr.nix [hunk 2]
```diff
@@ -29,7 +29,7 @@
 , postgresql
 , proj
 , protobuf
-, python3
+, python311
```
Comment (imincik) [inline pkgs/applications/gis/qgis/unwrapped-ltr.nix:32]: I am not really sure if we need to pin python in QGIS. [QGIS is building successfully](https://hydra.nixos.org/job/nixpkgs/trunk/qgis.x86_64-linux/all) even without this PR on Hydra.

File: pkgs/applications/gis/qgis/unwrapped.nix [hunk 3]
```diff
@@ -29,7 +29,7 @@
 , postgresql
 , proj
 , protobuf
-, python3
```
Comment (imincik) [inline pkgs/applications/gis/qgis/unwrapped.nix:32]: Not sure about pinning here either.

**Review by imincik — DISMISSED** [review]

File: pkgs/applications/gis/qgis/unwrapped.nix (hunk 3 above)
Comment (fpletz) [reply pkgs/applications/gis/qgis/unwrapped.nix:32]: But it fails to start. :smile:

**Comment by fpletz** [comment]
> [QGIS is building successfully](https://hydra.nixos.org/job/nixpkgs/trunk/qgis.x86_64-linux/all) even without this PR on Hydra.

Then it maybe was the gdal update? Not sure judging by hydra.

**Comment by fpletz** [comment]
@imincik I disagree because right now QGIS doesn't work at all. With this PR at least `qgis` works. I don't know why `qgis-ltr` fails but this PR improves the situation.

**Comment by fpletz** [comment]
The be more precise: Without this PR, `qgis` fails to evaluate:

```
error: wxpython-4.2.1 not supported for interpreter python3.12
```

With the grass changes `qgis` builds but fails to start.

With the gross and qgis changes from this PR, `qgis` works. `qgis-ltr` fails in the same way with or without the changes in this PR but the changes _are_ necessary until wxpython supports python 3.12.

**Comment by imincik** [comment]
>  but the changes are necessary until wxpython supports python 3.12.

I am not able to investigate now, but we have [python312Packages.wxpython](https://search.nixos.org/packages?channel=unstable&from=0&size=50&sort=relevance&type=packages&query=python312Packages.wxpython).

Will have a look ASAP.

**Comment by fpletz** [comment]
> I am not able to investigate now, but we have [python312Packages.wxpython](https://search.nixos.org/packages?channel=unstable&from=0&size=50&sort=relevance&type=packages&query=python312Packages.wxpython).

Which fails to evaluate. :smile_cat:

**Review by globin — APPROVED** [review]
The pins can be removed again when wxpython supports python 3.12. This is the logical solution for the time being.

**Comment by imincik** [comment]
@fpletz , @globin qgis-ltr should be fixed in https://github.com/NixOS/nixpkgs/pull/326550 . Please review if you can.

**Pull request merged**